src/api-errors.js

```diff
@@ -77,3 +77,44 @@ export const invalidParameters = create(10, (message, errors) => {
     message: message || 'invalid parameters'
   }
 })
+
+export const vmMissingPvDrivers = create(11, () => ({
+  message: 'missing PV drivers'
+}))
+
+export const vmIsTemplate = create(12, () => ({
+  message: 'VM is a default template'
+}))
+
+export const vmBadPowerState = create(13, params => ({
+  data: { powerState: params[2], requiredPowerState: params[1] },
+  message: `VM state is ${params[2]} but should be ${params[1]}.`
+}))
+
+export const vmLacksFeatureShutdown = create(14, () => ({
+  message: 'VM lacks feature shutdown'
+}))
+
+export const notSupportedDuringUpgrade = create(15, () => ({
+  message: 'not supported during upgrade'
+}))
+
+export const pifVlanExists = create(16, () => ({
+  message: 'PIF VLAN already exists'
+}))
+
+export const vdiInUse = create(17, () => ({
+  message: 'VDI in use'
+}))
+
+export const hostOffline = create(18, () => ({
+  message: 'host offline'
+}))
+
+export const operationBlocked = create(19, () => ({
+  message: 'operation blocked'
+}))
+
+export const patchPrecheckFailedIsoMounted = create(20, () => ({
+  message: 'patch precheck failed'
+}))
```

Comment:
Should probably be a bit more generic (https://github.com/xapi-project/xen-api/blob/master/ocaml/idl/datamodel.ml#L1172-L1185).

Maybe not, errors of the same type should be handleable homogeneously.